### PR TITLE
Issue/vivo 1908

### PR DESCRIPTION
**Pull request by brianjlowe**
**[VIVO-1908](https://jira.lyrasis.org/browse/VIVO-1908)**:

* Also addresses subtask https://jira.lyrasis.org/browse/VIVO-1906

# What does this pull request do?
Uses the existing language-filtering mechanism for generating dropdown select lists with language-appropriate labels on editing forms (such as "organizer of event") where the URIs are in the select lists are harcoded in the form.  Also adds a fallback mechanism to the manage webpages for individual form where language filtering is injected into a SPARQL query.

# What's new?
See GeneratorUtil.java and ManageWebapagesForIndividualGenerator.java for the most substantive changes.  
GeneratorUtil now has a method buildResourceAndLabelFieldOptions() that is very similar to the original method for building a list of constant field options.  This new method uses a WebappDaoFactory, if available, to retrieve in-memory class labels, and an RDFService to retrieve labels for all other URIs.  The individual editing forms supply the existing WebappDaoFactory and RDFService from the request, which already perform language filtering.  The editing forms no longer build DESCRIBE queries where language filtering is injected.

ManageWebpagesForIndividual.java adds fallback languages to the SPARQL query that retrieves webpage links as well as the labels for the link types.  Because this query involves GROUP BY, language filtering here cannot be handled by the RDFService-level filtering.

# How should this be tested?
* Create/navigate to a person.  From the Service tab, select "organizer of event".  Change the language selector, and note that event type labels in the dropdown continue to match the selected language.  For more in-depth testing, add a new URI to the list in AddOrganizerRoleToPerson.java, create that class in VIVO and supply a label in only one language, and note that the class continues to appear in the dropdown even if a label is not available that matches your current language selection.
* Go to the class hierarchy, select all classes, find "F1000 Link", and add a new subclass of that class.  Supply a label in only one language.  Navigate to a person.  Open the manage webpages page (Rolodex) and enter a new webpage of the new type you just created.  From the manage webpages page, change the language selector and note that the label for the new type continues to appear even if a label does not exist that matches the current language selection.

### CODE REVIEW

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ManageWebpagesForIndividualGenerator.java [hunk 1]
```diff
@@ -155,18 +155,36 @@ public EditConfigurationVTwo getEditConfiguration(VitroRequest vreq, HttpSession
...
-        + "    OPTIONAL { ?link core:rank ?rank } \n"
+        + "    OPTIONAL { ?link core:rank ?rank_ } \n"
         + "    OPTIONAL { ?link vitro:mostSpecificType ?type } \n"
-        + "    OPTIONAL { ?type rdfs:label ?typeLabel . \n"
-     // UQAM-Linguistic-Management Add linguistic control on label
-        + "                FILTER (lang(?typeLabel) = 'LANGUAGE' ) } \n"
-        + "} GROUP BY ?rank ?vcard ?link ?url ?typeLabel \n"
+        // UQAM-Linguistic-Management Add linguistic control on label
+        // Try full locale 
+        + "    OPTIONAL { ?type rdfs:label ?typeLabelPrimary . \n"
```
Comment (awoods) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ManageWebpagesForIndividualGenerator.java:170]: Nice!

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ManageWebpagesForIndividualGenerator.java [hunk 2]
```diff
@@ -213,10 +231,14 @@ protected String getEditForm() {
...
-         * UQAM-Linguistic-Management Adjust the query to the liguistic context
+         * UQAM-Linguistic-Management Adjust the query to the linguistic context
          */
-        Locale lang = SelectedLocale.getCurrentLocale(vreq);
-    	return WEBPAGE_QUERY.replaceAll("LANGUAGE", lang.toString());
+        Locale locale = SelectedLocale.getCurrentLocale(vreq);
+        ParameterizedSparqlString queryPstr = new ParameterizedSparqlString(
+                WEBPAGE_QUERY);
+        queryPstr.setLiteral("locale", locale.toString().replace("_", "-"));
+        queryPstr.setLiteral("language", locale.getLanguage().toString());
```
Comment (awoods) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ManageWebpagesForIndividualGenerator.java:240]: `toString()` is redundant here.

**Comment by awoods** [comment]
Some minor comments... functionality works as advertised.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddAttendeeRoleToPersonGenerator.java [hunk 3]
```diff
@@ -21,49 +22,28 @@ String getRoleType() {
     /** Editor role involves hard-coded options for the "right side" of the role or activity. */
```
[the remark this answers is not in the document]
Comment (brianjlowe) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddAttendeeRoleToPersonGenerator.java:22]: It actually refers to the individual who is not the bearer of the role, but good catch.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ManageWebpagesForIndividualGenerator.java (hunk 2 above)
Comment (brianjlowe) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ManageWebpagesForIndividualGenerator.java]: Removed.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddResearcherRoleToPersonGenerator.java [hunk 4]
```diff
@@ -23,24 +24,15 @@ public String getRoleType() {
...
-//		        "http://vivoweb.org/ontology/core#Grant", "Grant",
-//	        "http://purl.obolibrary.org/obo/ERO_0000015", "Human Study",
-//	        "http://vivoweb.org/ontology/core#Project", "Project",
-//	        "http://purl.obolibrary.org/obo/ERO_0000014", "Research Project");
+	    return GeneratorUtil.buildResourceAndLabelFieldOptions(
+                vreq.getRDFService(), vreq.getWebappDaoFactory(), "", 
+                I18n.bundle(vreq).text("select_type"),
+                "http://vivoweb.org/ontology/core#Grant",
+                "http://purl.obolibrary.org/obo/ERO_0000015" /* Human Study" */,
+                "http://vivoweb.org/ontology/core#Project", "Project",
```
[the remark this answers is not in the document]
Comment (brianjlowe) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddResearcherRoleToPersonGenerator.java]: Yep: now gone.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/GeneratorUtil.java [hunk 5]
```diff
@@ -2,81 +2,126 @@
...
+	   List<String> options = new ArrayList<String>();
+	   if(headerValue != null && headerLabel != null) {
+	       options.add(headerValue);
+	       options.add(headerLabel);
+	   }
+	   IRIFactory iriFactory = IRIFactory.iriImplementation();	   
+	   for(String resourceURI : resourceURIs) {	       
+	       IRI iri = iriFactory.create(resourceURI);
+	       if(iri.hasViolation(false)) {
+	           log.debug("Not adding invalid URI " + resourceURI
```
[the remark this answers is not in the document]
Comment (brianjlowe) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/GeneratorUtil.java]: Elevated to warn.

**Pull request merged**